src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -4356,6 +4356,71 @@ TEST_CASE("AdaptiveMeanPoolingTestCase", "[ANNLayerTest]")
   REQUIRE(arma::accu(delta) == 2.25);
 }
 
+/**
+ * Simple Separable Convolution layer test.
+ */
+TEST_CASE("SimpleSeparableConvolutionLayerTest", "[ANNLayerTest]")
+{
+  arma::mat input, output, delta;
+  // Input channels = 3, output channels = 21 and groups = 3
+  // Input dim = (224, 224)
+  SeparableConvolution<> module1(3, 21, 3, 3, 1, 1, 0, 0, 224, 224, 3);
+
+  // Test the forward function.
+  input.set_size(224 * 224 * 3, 1);
+  input.fill(2);
+  module1.Parameters().fill(1);
+  module1.Reset();
+  module1.Forward(input, output);
+
+  // Comparison value taken from PyTorch conv layer with the same config
+  REQUIRE(arma::accu(output) == 19664316);
+
+  // Test the backward function.
+  module1.Backward(input, output, delta);
+  // Calculated using nn.functional.conv2d 
+  REQUIRE(arma::accu(delta) == )
```

Comment:
I have left these blank as i didn't knew how to compute the backward value in PyTorch to compare.

Reply:
Hey, Could you take a look at this notebook for calculating values (https://colab.research.google.com/drive/1V-g_NmKMuM2EAm79CDwwAXswca1WbE0u?usp=sharing).

Reply:
So i wrote this 

```python3
import torch
import torch.nn as nn

conv = nn.Conv2d(1, 4, 3, 1, 0, 1, 1)

conv.weight = nn.Parameter(torch.ones_like(conv.weight))
conv.bias = nn.Parameter(torch.ones_like(conv.bias))
one = torch.ones((1, 1, 20, 20), requires_grad=True)
one = one * 2
one.retain_grad()
out = conv(one)
out.sum().backward(retain_graph=True)
print(torch.sum(one.grad))
```
and the output is  :

```bash
tensor(11664.)
```

While for mlpack i have this code : 

```c++
  arma::mat input, output, delta;
  SeparableConvolution<> module2(1, 4, 3, 3, 1, 1, 0, 0, 20, 20, 1);

  // Test the forward function.
  input.set_size(20 * 20, 1);
  input.fill(2);
  module2.Parameters().fill(1);
  module2.Reset();
  module2.Forward(input, output);

  // Comparison value taken from PyTorch conv layer with the same config
  std::cout << arma::accu(output) << std::endl; //24624

  module2.Backward(input, output, delta);
  std::cout << arma::accu(delta) << std::endl;
```

whose output is : 

```bash
24624   <- this being the forward output
221616 <- and this is the delta
```

So now i am unsure about two things: 
1. Am I comparing the right values ?
2. Is the code for both the tests right ?

Reply:
I think the comparison is correct.

Reply:
@zoq can you take a look too ?

Reply:
Can you print the shape of the gradient and weights as well.